Support for Websocket Transport Layer

brunombsilva opened the pull request:
For an internal project at Talkdesk we had the need to connect to a MQTT server via websockets in one of our Elixir projects.

This PR includes several features, but we tried to tidy up the commits so that you can eventually accept some that may make sense for Hulaaki, while still being able to refuse others.

- We added a `Hulaaki.Transport` behavior that is used as a contract to implement transport layers other than ssl and tcp.
- We've made some improvements around tests such as explicitly fail upon `Connection.connect` during setup, made the tests async to improve performance
- We created a new transport layers based on the `Hulaaki.Transport` behavior to support websockets. We made tried both [Socket](https://github.com/meh/elixir-socket) and [Gun](https://github.com/ninenines/gun) but opted to suggest Socket due to stability/compatibility concerns.
 - We promoted the library to Application to allow us to have a supervision tree that takes care of connection processes and deal with failures (by providing a new `on_connection_down` callback.
- While working on the project we tested and updated it to support Elixir >= 1.6 and OTP up to 21.

Commit:
Sometimes tests failed due to a setup error, those errors are now visible.

Commit:
Supporting OTP 21

Commit:
Abstract transport layer

Instead of explicitly using :ssl or :gen_tcp, we introduce a new abstraction: Hulaaki.Transport that allow us to implement other transport layers.

Commit:
Creating a new Transport to support MQTT via websocket.

Commit:
Creating an Application and supervision tree to better deal with process failure.

Also, a new callback was added: on_connection_down.
Elixir requirements bumped to 1.6 due to DynamicSupervisor support.

Commit:
In order to get faster tests, they are now async.

To achieve this we had to ensure that were no client_id name conflicts.
Almost all timeout values are now consistent.

Commit:
Update README with websockets and configurations changes

suvash commented:
@brunombsilva @ejscunha This looks great. I'm assuming the tests are passing locally. Kinda busy right now, so give me a week or two to go through this. Hopefully, we'll get most of these changes merged. 🍻

Commit:
Flowing transport/connection errors into the explicit client calls

Such as ping, subscribe, unsubscribe. This way if something is wrong with the connection you'll get an error tuple instead of always :ok

Commit:
Support supervised connections on Elixir 1.5 and 1.4

Since we were using a DynamicSupervisor, Elixir 1.5 and below were
not supported, so the supervisor was changed into a standard
supervisor, while trying to maintain the behaviour of a DynamicSupervior.

brunombsilva commented on README.md:
```diff
@@ -89,6 +89,22 @@ Please check the [CHANGELOG.md](https://github.com/suvash/hulaaki/blob/master/CH
 
 ## Documentation
```
@suvash , I'll start a thread here regarding usage samples. The actual usage of the library remains the same. Since Elixir 1.4 that the applications are inferred from the dependencies and booted with it's supervision tree.

Regarding actual usage, we still didn't include Hulaaki in our project. We will start working on that next week or so.

I've updated the README with more specific transport layer support and configuration, tough.

suvash replied:
Sounds good. I'll wait for a few weeks until you've been able to integrate this in your project (using your fork), just in case you run into any bugs etc.
Also, I haven't yet gone through the code, but I'm assuming that the Connection Genserver instance is now started by the Application > Client supervision tree. When integrating, can you also check if it's straightforward to open multiple connections (to multiple servers) with this Supervision tree structure ? 
After all looks good, we should be able to clean up and merge.

ejscunha replied:
Hey, first off all thank you @suvash for this great library! If you need to connect to multiple servers, you will need to have a client for each server, since the relation of the clients is 1:1 to connections (as it was before if I'm not wrong). The only guarantee that we will need is to have a different `client_id` for each client, which we use as id for each supervised connection.

suvash replied:
Yeah. I think the Client : Connection is 1:1 now and was before as well. I was thinking more about the Supervision tree that's started as a part of the application. If I've understood this correctly, the application supervision tree will start an instance of the Hulaaki tree (with client and connection genservers), so was thinking if that'd be straightforward to scale to multiple clients+connections as well.
However, I still haven't had the time to look at the code yet. Vacation around the corner and been busy. If I don't get the time this week, I'll most likely only be able to look at towards the end of August.
But, that should be alright, and perhaps enough time for you to integrate it in your Elixir project and test run in production(?). Looking forward to the integration news.
And yeah, thanks a lot for this PR @brunombsilva @ejscunha, Lots of good stuff here that I've been planning to implement for a while but hadn't had the time to. :beers:

suvash replied:
So, how has it gone with the integration ?

ejscunha replied:
Started working on it just now :)
When we get something working we will give news.

ejscunha replied:
Hello again, sorry for all this time without saying anything. We started using Hulaaki with websockets in our QA environment to make tests, it looks to be working ok. Sometimes the connection goes down for some reason, maybe due to some issue with the keep alive option but I still need to investigate this further.

In our use case we need to open many connections to a MQTT server, since we can't re use the same connection for different topics, so while trying to include the clients in our supervision tree it doesn't feel natural and I think it needs to be improved.

We setup a DynamicSupervisor responsible for supervising all the new clients on demand and to start the clients we have a task that requests the supervisor to start each client.

Something like this:

``` elixir
```

ejscunha replied:
Hello again, sorry for all this time without saying anything. We started using Hulaaki with websockets in our QA environment to make tests, it looks to be working ok. Sometimes the connection goes down for some reason, maybe due to some issue with the keep alive option but I still need to investigate this further.

In our use case we need to open many connections to a MQTT server, since we can't re use the same connection for different topics, so while trying to include the clients in our supervision tree it doesn't feel natural and I think it needs to be improved.

We setup a DynamicSupervisor responsible for supervising all the new clients on demand and to start the clients we have a task that requests the supervisor to start each client.

Something like this:

``` elixir
defmodule Bootstrapper do
  use Task
  alias Clients.Supervisor

  def start_link(_) do
    Task.start_link(__MODULE__, :run, [])
  end

  def run do
      clients = [..]

      Enum.each(clients, fn client ->
        Supervisor.start_client(client)
      end)
  end
end

defmodule Clients.Supervisor do
  use DynamicSupervisor

  def start_link(_) do
    DynamicSupervisor.start_link(__MODULE__, :ok, name: __MODULE__)
  end

  def init(:ok) do
    DynamicSupervisor.init(strategy: :one_for_one)
  end

  def start_client(client) do
    child_spec = %{
      id: client,
      start: {Client, :start_link, [:start, client]},
      restart: :transient
    }

    DynamicSupervisor.start_child(__MODULE__, child_spec)
  end
end

defmodule Client do
  use Hulaaki.Client

  def start_link(:start, client) do
    case __MODULE__.start_link(%{}) do
      {:ok, pid} ->
        send(pid(, {:connect_and_subscribe, client})
        {:ok, pid}

      other ->
        other
    end
  end

  def handle_info({:connect_and_subscribe, client}, state) do
    Task.start(__MODULE__, :connect_and_subscribe, [client, self()])

    {:noreply, state}
  end

  def connect_and_subscribe(client, pid) do
    with {:ok, topic} <- connect({client, pid}),
            :ok <- subscribe(pid, topics: [topic], qoses: [1]) do
      :ok
    else
      error -> error
    end
  end

  defp connect({client, pid}) do
    ..
    opts = [..]

    case __MODULE__.connect(pid, opts) do
     end
  end
end
```

suvash replied:
Hi again @ejscunha , Thanks for the feedback above. I was sorta expecting issues around the supervisor, it's kinda tricky to setup one type of tree that might work for all end users of the library.
I'm planning to incorporate these changes into the library sometime in October. Do you have any new insights/feedback that I could make use of, based on any further changes you've had to do ?

Commit:
Fix connection processes leak

There was an issue with connection processes that were being leaked
when there was connection problems (timeouts, disconnection, etc),
connection processes were being wrongly restarted when they should
be stopped.
